[enriched-gitlabcomments] New enricher to handle gitlab comments

vchrombie opened the pull request:
This PR proposes a new gitlab enricher to handle gitlab comments. The old enricher is left in place to avoid breaking changes with existing customer dashboards.

Ref: https://github.com/chaoss/grimoirelab/issues/208

vchrombie commented:
The PR has a lot of pending work, I will be working on this. I will mark to ready for review after completing some good work.

Please let me know if you are having any suggestions.

valeriocos commented:
Hi @vchrombie , please ping me when I can review this PR. I had a look at the current status and the start looks promising :)

WRT tests (if this can be of any help), you should run Perceval and copy some docs to a test file, which will be finally put in the folder [tests/data](https://github.com/chaoss/grimoirelab-elk/blob/master/tests/data/). The name is important since it is used to automatically identify the file and load its content (https://github.com/chaoss/grimoirelab-elk/blob/master/tests/base.py#L137)

If needed you can tweak the test data (e.g., changing value attributes) to make sure that all code is covered when running the tests.

vchrombie commented:
Hi @valeriocos 
I have completed the major part of the PR. I am yet to complete the `__get_reactions` function and I am working on that right now.

valeriocos commented:
Thanks for the update @vchrombie

vchrombie commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,586 @@
+        rich_mr['gitlab_repo'] = rich_mr['repository'].replace(GITLAB, '')
+        rich_mr['gitlab_repo'] = re.sub('.git$', '', rich_mr['gitlab_repo'])
+
+        # GMD code development metrics
```
I was building this enricher on top of `github2`.

GMD metrics implemented in _github2_ are
- number of forks (afaik, this is not possible in gitlab as we don't have the number of forks data)
- code merge duration (implemented :+1:)
- number of review comments (not done, but can be implemented by taking the number of `notes_data` of the MR WDYT?)

GMD metrics implemented in _gitlab_ are
- code merge duration (implemented :+1:)
- number of versions (done, will this be good to go here WDYT?)

I would like to know your comments on this. Let me know if I have to provide some perceval data to analyze.

valeriocos replied:
> number of review comments (not done, but can be implemented by taking the number of notes_data of the MR WDYT?)

Perfect, agree on using notes_data

> number of versions (done, will this be good to go here WDYT?)

Yes, it makes sense to have it in the enriched index

vchrombie replied:
Awesome, I will update the PR.

vchrombie replied:
Done. :+1:

vchrombie commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,586 @@
+            # Copy data from the raw comment
+            ecomment['body'] = comment['body'][:self.KEYWORD_MAX_LENGTH]
+            ecomment['body_analyzed'] = comment['body']
+            # ecomment['url'] = comment['html_url']
```
I think we have to remove this as we don't have any comment URL in the data like in _github_.

valeriocos replied:
Can we build a url with the information we have in the document? These URLs are good to allow navigating from the dashboard to the upstream item

vchrombie replied:
Let me check and get back to you. :runner:

vchrombie replied:
Hi @valeriocos 
I think we can reproduce the URL to these comments.

I am taking this MR as example, https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36.

In gitlab, every event is registered in the `notes_data` such as approving, commenting, merging, etc. I think we can consider all of them as they are provided with the API. I will address these events as comments itself.

So, each comment has an `id`. So we can build the URL using that `id` like `https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36#note_id`.

I have provided some examples from the API data itself.

-  @yashk2000 approved this merge request : https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36#note_325524652
-  Try formatting it properly as it is taking more characters in single line. Maybe from android studio you can use Ctl+Alt+L. : https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36#note_325525231
-  @immadisairaj approved this merge request : https://gitlab.com/amfoss/cms-mobile/-/merge_requests/36#note_325529463

TL;DR 
It is feasible to construct the URL for comment using the `id`

vchrombie replied:
Can you check it once and let me know how it looks?

valeriocos replied:
Thank you, I have checked this while testing the connectors

vchrombie commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,586 @@
+            # Copy data from the raw comment
+            ecomment['body'] = comment['body'][:self.KEYWORD_MAX_LENGTH]
+            ecomment['body_analyzed'] = comment['body']
+            # ecomment['url'] = comment['html_url']
```
same as above comment, think we need to remove this

valeriocos replied:
Same comment as above

vchrombie replied:
sam as above :D

vchrombie commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,586 @@
+
+COMMENT_TYPE = 'comment'
+ISSUE_COMMENT_TYPE = 'issue_comment'
+REVIEW_COMMENT_TYPE = 'review_comment'
```
Do you think we need to change this to 
```
MERGE_COMMENT_TYPE = 'merge_comment'
```
?

valeriocos replied:
Yes, it is clear, thanks

vchrombie replied:
Done. :+1:

vchrombie commented:
I will mark it ready-for-review once I complete the `__get_reactions` function too.

vchrombie commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,602 @@
+
+        return None
+
+    def __get_reactions(self, item):
```
Hi @valeriocos 

I have completed implementing the `__get_reactions` function. 

gitlab reactions data is fetched a bit different than github. 
You can see the format here, [link](https://gist.github.com/vchrombie/a8b3449c924ccd01948d1a64d1d93be5).

The output of the current implementation looks something like this :point_down: 
```
{
    'reaction_thumbsup': 4, 
    'reaction_basketball': 2, 
    'reaction_thumbsdown': 1, 
    'reaction_rocket': 1, 
    'reaction_8ball': 1
}
```

valeriocos replied:
Since GitLab provides far many emojis than GitHub, I would suggest to add an attribute that counts the total number of reactions. To avod having too many mappings related to emojis (with the current implementation we will have a mapping for `reaction_thumbsup`, one for `reaction_8ball`, etc.), I would suggest to have just one mapping that contains the name/type of the emoji and the corresponding count (see example below). WDYT?
```
"reaction" : {
            "properties" : {
               "count" : {
                "type" : "long"
              },
               "type" : {
                "type" : "keyword"
              }
            }
          },
```

valeriocos replied:
current mappings:
```
          "reaction_100" : {
            "type" : "long"
          },
          "reaction_basketball_player" : {
            "type" : "long"
          },
          "reaction_bow" : {
            "type" : "long"
          },
          "reaction_bow_tone2" : {
            "type" : "long"
          },
          "reaction_bug" : {
            "type" : "long"
          },
          "reaction_call_me" : {
            "type" : "long"
          },
          "reaction_call_me_tone2" : {
            "type" : "long"
          },
          "reaction_cartwheel_tone3" : {
            "type" : "long"
          },
          "reaction_cartwheel_tone5" : {
            "type" : "long"
          },
          "reaction_clap" : {
            "type" : "long"
          },
          "reaction_clap_tone3" : {
            "type" : "long"
          },
          "reaction_confetti_ball" : {
            "type" : "long"
          },
          "reaction_cry" : {
            "type" : "long"
          },
          "reaction_eyes" : {
            "type" : "long"
          },
          "reaction_golfer" : {
            "type" : "long"
          },
          "reaction_grinning" : {
            "type" : "long"
          },
          "reaction_heart" : {
            "type" : "long"
          },
          "reaction_heart_eyes" : {
            "type" : "long"
          },
          "reaction_hearts" : {
            "type" : "long"
          },
          "reaction_heavy_plus_sign" : {
            "type" : "long"
          },
          "reaction_hugging" : {
            "type" : "long"
          },
          "reaction_joy" : {
            "type" : "long"
          },
          "reaction_kissing_smiling_eyes" : {
            "type" : "long"
....
```

vchrombie replied:
I have corrected the implementation. Thanks for the help. :smile:

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,602 @@
+        if user is not None and user:
+            rich_issue['user_name'] = user['name']
+            rich_issue['author_name'] = user['name']
+            rich_issue["user_domain"] = self.get_email_domain(user['email']) if user['email'] else None
```
```
-            rich_issue["user_domain"] = self.get_email_domain(user['email']) if user['email'] else None
+            rich_issue["user_domain"] = self.get_email_domain(user['email']) if user.get('email', None) else None
```
Emails are currently not exposed by the GitLab API (AFAIK). We can keep the code, but we should make sure that `get_email_domain` is called only when `user['email']` exists.

vchrombie replied:
Yes, this seems more better. :+1:

vchrombie replied:
I have updated the same for `assignee` and `merged_by` data too.

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,602 @@
+            ecomment['body'] = comment['body'][:self.KEYWORD_MAX_LENGTH]
+            ecomment['body_analyzed'] = comment['body']
+
+            ecomment['url'] = eitem['issue_url'] + '#note_' + comment['id']
```
```
-            ecomment['url'] = eitem['issue_url'] + '#note_' + comment['id']
+            ecomment['url'] = eitem['issue_url'] + '#note_' + str(comment['id'])
```

comment['id'] should be casted to str. We could use `format`, as done for calculating `ecomment['id']`

vchrombie replied:
Perfect! :+1:

vchrombie commented:
> https://github.com/chaoss/grimoirelab-elk/pull/881#pullrequestreview-425802955

Thanks for the review @valeriocos.
I have just updated the PR with the changes you have suggested.

> Can you add the tests and schema?

Sure, I will be working on it and add them here soon.

vchrombie commented:
Recent updates to the PR
1. I removed the reactions mapping as we discussed it is not needed and also implemented the new `__get-reactions()` function.
2. added the schema using the gitlabcomments-mrs/issues_enriched index.
3. added the tests, (highly need a review on this as I'm not really good with it).

vchrombie commented on schema/gitlabcomments_issues.csv:
```diff
@@ -11,10 +11,10 @@ assignee_org_name,keyword,true,"Assignee organization name."
 assignee_user_name,keyword,true,"Assignee user name from SortingHat."
 assignee_uuid,keyword,true,"Assignee UUID from SortingHat."
 author_bot,boolean,true,"True if the given author is identified as a bot, from SortingHat profile."
-author_domain,keyword,true,"Domain associated to the author in SortingHat profile."
```
Hi @valeriocos 
I was confused about this deletion.

I didn't see this (`author_domain`) in the mapping too, but `assignee_domain` and `merged_by_domain` are present.
Can you check test the enricher once and let me know if you have this?

The same deletion is present in the merge_request too.

valeriocos replied:
`*_domain` fields are added by SortingHat. By default GitLab users don't have their emails exposed on the API. However, if you execute sortinghat with a matching algorithm that joins the identities based on username and email (the algorithms are available [here](https://github.com/chaoss/grimoirelab-sortinghat/tree/master/sortinghat/matching)), the email information can be collected. The `*_domain` fields should be filled and thus they will appear in the mapping.

vchrombie replied:
Okay, got it. So, I will add them back to the schema then.

valeriocos replied:
Perfect, let me know when I can review the PR, thanks!

vchrombie commented:
Also, I have made a small script to generate schema file of an es index, [generate-es-index-schema-py](https://gist.github.com/vchrombie/bf6a682edcf47624126317897e58679c#file-generate-es-index-schema-py). 
This was built on top of @animeshk08's script as it had some backlogs. 

The script can be improved by adding the idea of default field and adding descriptions. Please leave your comments in the gist or you can open an issue here, [vchrombie/grimoirelab-scripts](https://github.com/vchrombie/grimoirelab-scripts). Feedback is highly welcomed. ^^

vchrombie commented:
Hi @valeriocos 
I have updated the PR as per your suggestions.
Can you review this when you are free?
Thanks.

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+        reaction_dates.extend(comments_dates)
+
+        if reaction_dates:
+            return min(reaction_dates)
```
To increase the test coverage use a single return. You can define a variable (e.g., `first_attention`), which is set to None and modified in this if statement

vchrombie replied:
Nice hack.

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+        if review_dates:
+            return min(review_dates)
+
+        return None
```
Similar comment as above, please use just a return

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+            rich_issue['author_name'] = author['name']
+            rich_issue['author_domain'] = self.get_email_domain(author['email']) if author.get('email') else None
+        else:
+            rich_issue['author_login'] = None
```
This else block isn't covered by the test data

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+            rich_issue.update(self.get_item_project(rich_issue))
+
+        if 'project' in item:
+            rich_issue['project'] = item['project']
```
Not covered by the tests

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+            rich_mr['author_name'] = author['name']
+            rich_mr['author_domain'] = self.get_email_domain(author['email']) if author.get('email') else None
+        else:
+            rich_mr['author_login'] = None
```
Not covered by the test data

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+            rich_mr.update(self.get_item_project(rich_mr))
+
+        if 'project' in item:
+            rich_mr['project'] = item['project']
```
Not covered by the tests

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+        elif item['category'] == 'merge_request':
+            rich_item = self.__get_rich_merge(item)
+        else:
+            logger.error("[gitlab] rich item not defined for GitLab category {}".format(
```
Not covered by the tests

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+            ins_items += self.elastic.bulk_upload(items_to_enrich, self.get_field_unique_id())
+
+        if num_items != ins_items:
+            missing = num_items - ins_items
```
Not covered by the test data

valeriocos commented on grimoire_elk/enriched/gitlabcomments.py:
```diff
@@ -0,0 +1,573 @@
+            if len(items_to_enrich) < MAX_SIZE_BULK_ENRICHED_ITEMS:
+                continue
+
+            num_items += len(items_to_enrich)
```
Not covered by the tests

valeriocos commented:
please @vchrombie ping me when the PR is ready for review, thanks

almereyda commented:
Coming from outside and having dismissed GrimoireLab previously for lacking GitLab support, I would still like to see it support additional code forges other than GitHub, as per the arguments in inhttps://github.com/chaoss/grimoirelab/issues/284#issuecomment-827155190).

Is there something me as an outsider can contribute to the formal aspects of the PR, in so the remaining technical questions can be answered more easily? Speaking of:

- [ ] rebase on master
- [ ] update elk readme
- [ ] add mordred supported data sources documentation (PR)
- [ ] update tests for edge cases and error conditions

vchrombie commented:
Hi @almereyda, thanks for your comment.

> Coming from outside and having dismissed GrimoireLab previously for lacking GitLab support

Sorry about it, but the current GrimoireLab supports analyzing issues and merge requests from the gitlab projects ([reference](https://github.com/chaoss/grimoirelab-sirmordred#gitlab-)). This PR is an additional enricher which can analyze the comments of the gitlab issues and mrs. I couldn't complete it because I completely forgot about this.

> Is there something me as an outsider can contribute to the formal aspects of the PR, in so the remaining technical questions can be answered more easily? Speaking of:

Thanks for offering help, I appreciate it. But, I think I can complete the work on this PR. I can pull some time during the coming weekend and complete te pending work.
